website/blog/2022-07-26-configuring-grants.md

```diff
@@ -0,0 +1,211 @@
+      For every role that can access at least one object in a schema,
+      grant 'usage' on that schema to the role.
+      That way, users with the role can run metadata queries showing objects
+      in that schema (a common need for BI tools)
+    */
+    {% set schema_grants = {} %}
+    {% if execute %}
+      {% for node in graph.nodes.values() %}
+        {% set grants = node.config.get('grants') %}
+        {% set select_roles = grants['select'] if grants else [] %}
+        {% if select_roles %}
+          {% set database_schema = node.database ~ "." ~ node.schema %}
+          {% if database_schema in database_schemas %}
+            {% do schema_grants[database_schema].add(select_roles) %}
+          {% else %}
+            {% do schema_grants.update({database_schema: set(select_roles)}) %}
+          {% endif %}
+        {% endif %}
+      {% endfor %}
+    {% endif %}
+    {% set grant_list %}
+      {% for schema in schema_grants %}
+        {% for role in schema_grants[schema] %}
+          grant usage on schema {{ schema }} to {{ role }};
+        {% endfor %}
+      {% endfor %}
+    {% endset %}
+    {{ return(grant_list) }}
+{% endmacro %}
+```
+
+This might be too clever -- but you get the idea!
+
+You can even do this at the *start* of the run, right after dbt creates its schemas, rather than waiting until the end. (Although it’s not a huge deal to wait.)
+
+```yaml
+on-run-start:
+	- {{ grant_usage_on_schemas_where_select() }}
+```
+
+### Advanced permissions (or other operations)
+
+Want to restrict access to specific rows in a table for specific users? Or dynamically mask column values depending on who’s asking?
+
+The approach varies by database: in Snowflake, you’ll still want a `post-hook` to apply a [row access policy](https://docs.snowflake.com/en/user-guide/security-row-intro.html) or a column [masking policy](https://docs.snowflake.com/en/sql-reference/sql/create-masking-policy.html) to your table whereas in Databricks you'd use [dynamic view functions](https://docs.databricks.com/security/access-control/table-acls/object-privileges.html#dynamic-view-functions).
+
+It’s good to have hooks and operations as a method to utilize cutting-edge database capabilities. Any cases that become a wide and clearly demonstrated need can be upgraded by being built into `dbt-core`.
+
+## Appendix
+
+### Caveats and trade-offs of the original guidance
+`on-run-end` hooks:
+> for the period of time between when a model runs, and the end of the run, no one will be able to query that model, instead they’ll get a “permission denied” error. This creates downtime in your BI tool.”
+
+`manage grants` privilege:
+> It is worth noting that this privilege *is* a global privilege – now anyone using the `transformer` role can change grants on any object as though they are the owner of the object. Up to you if you’re comfortable with this! If not, you may want to use a combination of `post-hooks` and `on-run-end` hooks instead 🙂”
```

Comment:
👍 good call including this, without centering it too much

(TIL that the British spell this "centring")